readme.md
# **circulaRprint**

**What is circularprint?**  
CirculaRprint is a controller software for circular plotters.
The plotter hardware consists of a rotating table and of an arm that moves across it in and arc. 
This eliminates the need for linear bearings/rods on the two main axis.
The tool tip (i.e. a pen, laser, whatever...) needs to be exactly centered above the table or else there will be a 
dead spot in the middle of the table that can't be reached.
A disadvantage of this approach is the inconsistent resolution, as it gets lower the further you get away from the 
table center (only applies to the table axis, not the arm).

Here are some images that should give you a basic understanding of how the plotter works:  
https://drive.google.com/open?id=1_3LzlJe7sTjrp4OGNJY0WJFvSQl1Wn2D  
or watch my youtube video:  
https://youtu.be/Dlr2wKOdKks

Feel free to contact me if you've got any questions:  
oehrly.dev@gmail.com
    
**Hardware**  
A Raspberry Pi is used as a controller board.  
Also a few H-Bridge motor drivers are needed.  
Of course some stepper motors (I took mine from old DVD drives and printers) and a few more components to build the rest from.

**Features**
* Read Gcode from File (obviously)
* Linear Movement G00/G01 (both treated the same)
* Circular Movement G02/G03
* Individual Axis acceleration and speed control (inspired by GRBL; thank you)
* Backlash correction

**FYI**
The table axis is also called Alpha or A Axis  
The arm is also referred to as beta arm, Beta axis or B Axis

## Configuration
The configuration is a simple python dictionary. Every line needs to end with a comma
(Except for comments behind it). The code mostly does not check whether the configuration is valid,
so if you enter invalid values it'll either crash or do something funny.

#### Explanation of the configuration:

**Key** | **Values** | **Explanation**
--------|------------|----------------
MA, MB, MZ | `[1, 2, 3, 4]` | GPIO pins for alpha, beta and z-axis stepper
M*_full_step_seq | `0` or `1` | Use full step sequence for more torque; results in half the resolution
M*_dir | `1` or `-1` | Changes the direction of the motor; same could be achieved by changing the pin order but this is easier
distortion_correction | `1` or `-1` | Changes whether an offset is added or subtracted during path calculation. For some combinations of motor directions this is necessary. It can't be corrected automatically as it also depends on how the motor is wired up. If straight lines are curved you should try changing this value.
M*_backlash_correction | `[steps, delay]` | Compensates for backlash when the motor/axis changes direction. The first number specifies how many extra steps should be taken. The second number specifies the delay per step. The delay should be approximately (1 divided by axis feed rate). A too low number might result in skipped steps. A too high number might cause non-smooth runs as this extra time is not taken into account when calculating the timing.    
M*_always_unhold | `False` or `True` | If set to True the motor coils are turned off immediately after each movement. This prevents the coils from heating up to much while the motor is holding position. The end position may not be reached correctly though and the motor has no active holding force. Only recommended for slow moving axes that don't move often aka the z-axis. 
rB | `length in mm` | The length of the beta arm in **millimetres** from its point of rotation to the tool tip. Should be measured as accurate as possible. If the tool is offset to the side of the arm, measure from tool tip to pivot point!
alpha_ /beta_res | `resolution in degree/step` | The resolution of the alpha and beta axes. Has to be given in **degree per step**! 
feed_rate | `speed in mm/s` | The nominal speed ot the tool relative to the table
acceleration | `acceleration in mm/s²` | Maximum acceleration of the tool relative to the table
minimum_feed_rate | `speed in mm/s` | Slowest speed for the tool to move at (relative to the table)
junction_deviation | `default: 0.01` | The higher this number is, the faster cornering speeds are. Start out slow!
a_/b_feed_rate | `speed in steps/s` | Maximum absolute speed for this axis.
a_/b_acceleration | `accleration in steps/s² ` | Maximum absolute acceleration for this axis 
z_feed_rate | `speed in mm/s` | Z-axis feed rate; no acceleration is available for z-axis
minimum_axis_feed_rate | `speed in steps/s` | Minimum speed for both alpha and beta axis
SegAP_buffer_length | `number of segments` | How many segments should be used for one calculation run of the SegmentAccelerationPlanner. If the number is too low, maximum speed is never reached. But especially for threaded calculation it should not be too high or it'll take too much time per calculation while doing less calculations overall. Changing this number won't change much about the overall time needed for calculation.
Default_mode | `0` or `1` | Mode 0 calculates all path and acceleration data before starting; this can easily take a minute for larger files. Mode 1 does the calculations in a separate thread while running; the higher CPU load might cause worse timing.  

The acceleration planner always ensures that no speed or acceleration limit is exceeded.  
This means in turn that not every axis is always running at the specified maximum as that isn't possible.


## How to run it
```python
from rprint import Controller

controller = Controller(mode=0, config={})

controller.load_gcode_file('example.ngc')

controller.run()

controller.move_home()
controller.shutdown()
```
This is the basic script for running the circulaRprint controller. It is basically the same as in run.py.  
`mode` and `config` (dict) can be provided optionally, else the defaults (config.py) are used.

I recommend running move_home(). It should not be necessary if the gcode ends at x0 y0, but just to be sure...  
shutdown() finally turns everything off


## Calibration 
#### Resolution
To determine the axis resolution I recommend you to create a python file with the following code:
```python
from rprint import Controller

controller = Controller()

controller.MB.move(100, 0.01)
controller.MA.move(100, 0.01)
controller.MB.move(-100, 0.01)
controller.MA.move(-100, 0.01)

controller.shutdown()
```
`move(steps, delay per step)`  

This should draw a triangle with curved sides. Two sides are drawn by the beta axis and one by the alpha axis.
Increase the number of steps moved in the code above, until the triangle is as big as possible. This makes measuring easier.
Now you need to measure the distance from one corner to the next one.
The following formula helps you to calculate how many degrees the axis moved:

a = arcsin(s /(2*r)) * 2

a = angle  
s = measured distance  

r is the radius of the circle
* for the alpha axis use the distance of the side drawn by it's movement to the centre 
* for the beta axis use the length of the arm as radius r

When you've calculated the angle for each axis, divide it by the number of steps you set in the code. 
Now you have the resolution in degree per step.    

#### Centering
Getting the pen tip centered on the table can be quite tricky. The only thing I can recommend is to lower the pen onto the table.
Then move the table half a rotation by hand. If you can only see a dot, everything is perfect.
If you can see half a circle, the tip is not centered. From the orientation of the circle and the direction you turned the table in, you can conclude the direction you're off in.
Move the position of the tip a bit and try again until no circle is drawn anymore.

## GCode
I only tested gcode produced by the GCode-Tools extension for Inkscape, different gcode might work.
There are a few things one should know about:
1. The controller expects the Z-axis to be at Z=0 when starting (a pen tip would need to be touching the table)
1. A circle segment may cover 180° of a circle at max
1. G00 and G01 are treated exactly the same (the speed my machine was working at was already the maximum my tiny steppers could handle. So I didn't bother to implement a difference.)
1. Feed rates specified in GCode are ignored
1. The code only executes changes of the z-height if there is no x and y information given in the same line (i.e. Z would be ignored in the following command: G00 X5 Y6 Z3 )

GCode should be in millimetres.

It ***might*** work in inch or different units if:
* the length of the arm is specified in inch instead mm
* speeds are given in in inch/s instead mm/s
* acceleration is given in inch/s² instead mm/s²

If it works please contact me so I can update this.

## DebugOut Tool

This is really only intended for debugging. It might crash your system if you draw large files. (It'll remind you)  
! Requires kivy installed ! 
  
To run it, you just start debugout.py  
It will then import the controller and pass an extended version of the default configuration to it.

Configuration of debugout is done directly in debugout.py.  

DebugOut can only be used to verify that the PathPlanner class is working correctly. If anything after that messes up the movement, you won't see it here.


changelog.txt
Changelog of RPrint
###############

v0.0.7:
- added source/provider chain:
    every class requests single PathSegments from its "provider"-class
    PathPlanner requests GCommands from a source and creates PathSegments from that
    serve_next() is a required function for all classes in the calculation chain
    advantages: Not everything needs to be precalculated
                Possibility of modular calculation chain
- added SegmentBuffers and the ability to use simultaneous thread calculation or pre-calculation
- removed accelerate_b_light in StepAP

v0.0.6:
- added Z axis movement
- added always unhold option

v0.0.5:
- correction of motor direction is now done in BipolarStepperMotor class
- huge improvments to SegmentAccelerationPlanner speed; now only takes 20%-30% of calculation time compared to v0.0.4
- improvements to minimum step delay in BipolarStepperMotorClass; now based on Motor max speed
- added distortion correction option in config
- fixed bug in SegmentAccelerationPlanner that cause unnecessary deceleration and acceleration
- code cleanup and improved naming conventions
- fixed showgcode option in debugout

v0.0.4:
- various changes to the GCommand and PathSegment data structure
- fixed issues with the calculation of the difference between old and new alpha values
    (this lead to the table turning in the wrong direction)
- added possibility to cancel print by using ctrl+c
- added Grbl inspired acceleration control featuring overall and per axis speed and acceleration limits
- added backlash correction

v0.0.3:
- replaced Flow data handling
- added disclaimer for debug out
- fixed no circle drawn if circle starts at origin
- improved code style in PathPlanner
- fixed calculation of gamma
- fixed no circle drawn if circle center is the same as table center

v0.0.2:
- made linear interpolation work
- switching PathPlanner linear calculations to vectors
- added config file
- added gcode parser + possibility of reading gcode from file
- introducing Flow for data processing
- added graphical DEBUG (only) output
- added arc interpolation (G02/G03) !!circle segments may be 180 degrees at maximum

v0.0.1:
 ++ first test ++







config.py
# circulaRprint                                 #
# Circular Plotter controller for Raspberry Pi  #
# Author: Oehrly, 2018                          #
#################################################
#
# configuration file
# make a backup of this file in case you do something wrong
#
# this is the config file for my hardware and it should give you a basic idea about how stuff works
# you will need to modify it to suit your setup

config = {

    'MA': [11, 9, 10, 22],  # Motor Alpha Pins
    'MB': [15, 14, 18, 23],  # Motor Beta pins
    'MZ': [8, 7, 25, 24],  # Motor Z Pins

    'MA_full_step_seq': 0,  # use full step sequence instead of default half step sequence
    'MB_full_step_seq': 0,  # ~ results in more torque but half the resolution
    'MZ_full_step_seq': 1,  # 1 = Full Step, 0 = Half Step

    'MA_dir': 1,  # change motor dir; same as switching cables around or changing pin order
    'MB_dir': 1,  # just provided for convenience reasons
    'MZ_dir': -1,  # hast to be either 1 or -1

    'distortion_correction': 1,  # either 1 or -1; necessary for some combinations of motor directions
    # (you'll notice; if straight lines ain't straight this should fix it)

    'MA_backlash_correction': [2, 0.003],  # extra steps to correct for backlash when changing direction
    'MB_backlash_correction': [5, 0.003],  # [number of steps, delay per step]
    'MZ_backlash_correction': [0, 0],

    'MA_always_unhold': False,  # unhold immediately after movement to prevent overheating of motor coils
    'MB_always_unhold': False,  # mostly useful on zaxis as it sometimes holds position for a long time
    'MZ_always_unhold': True,   # will cause issues with fast moving steppers as they might be turned off too soon and
                                # might miss steps

    'rB': 100,  # arm length  (mm)

    'table_offset': 0,  # (degree); changes starting position of table --> rotates everything by the angle specified

    'alpha_res': 0.0862,  # degree per step (table)
    'beta_res': 0.04182,  # degree per step (arm)
    'z_res': 0.001881,  # mm per step (z-axis)

    'feed_rate': 30,  # nominal head feed rate (mm/s)
    'acceleration': 20,  # maximum head acceleration (mm/sec^2)
    'minimum_feed_rate': 1,  # slowest speed for the head to move at (mm/s)
    'junction_deviation': 0.01,  # higher numbers result in faster cornering speeds

    'a_feed_rate': 400,  # maximum feed rate for A axis (steps per second)
    'a_acceleration': 150,  # maximum A axis acceleration (steps/sec^2)
    'b_feed_rate': 350,
    'b_acceleration': 150,
    'z_feed_rate': 200,  # z is driven without acceleration currently

    'minimum_axis_feed_rate': 50,  # used in acceleration planning for a and b axis (steps/sec)

    'SegAP_buffer_length': 200,  # length of the buffer holding segments currently being calculated in SegAP

    'Default_mode': 0,  # 0=pre-calculation, 1=simultaneous threaded calculation
}
